.ci/generate_bear_requirements.py

```diff
@@ -95,28 +102,26 @@ def get_inherited_requirements():
     return inherited_requirements
 
 
+def helper(requirements, instance_dict):
+    for requirement in requirements:
+        if isinstance(requirement, AnyOneOfRequirements):
+            helper(requirement.requirements, instance_dict)
+        elif any(type(requirement).__name__ == instance
+                 for instance in SUPPORTED_INSTANCES):
+            instance_dict[type(requirement).__name__].add(requirement)
+
+
 def get_all_requirements(bears):
-    pip_requirements = []
-    npm_requirements = []
-    gem_requirements = []
+    instance_dict ={}
+
+    for instance in SUPPORTED_INSTANCES:
+        instance_dict[instance] = set()
 
     for bear in bears:
-        for requirement in bear.REQUIREMENTS:
-            if isinstance(requirement, PipRequirement) and \
-               requirement not in pip_requirements:
-                pip_requirements.append(requirement)
-            elif isinstance(requirement, NpmRequirement) and \
-               requirement not in npm_requirements:
-                npm_requirements.append(requirement)
-            elif isinstance(requirement, GemRequirement) and \
-               requirement not in gem_requirements:
-                gem_requirements.append(requirement)
-
-    return (
-        sorted(pip_requirements, key=lambda requirement: requirement.package),
-        sorted(npm_requirements, key=lambda requirement: requirement.package),
-        sorted(gem_requirements, key=lambda requirement: requirement.package)
-        )
+        helper(bear.REQUIREMENTS, instance_dict)
+
+    return [sorted(instance_dict[instance], key=lambda requirement: requirement.package)
```

Comment:
80 cpl

this file isnt under CI yet, so check your changes manually with pyflakes, pycodestyle and pydocstyle.

but ... if you return `instance_dict` here, the sorting can happen outside of the function, and we avoid returning tuples which are inherently unsuitable here as the size of them increases as new requirement types are added.

Reply:
done checking under coala